src/qutip_qip/decompose/_utility.py

```diff
@@ -29,3 +30,166 @@ def check_gate(gate, num_qubits):
+
+    # array index for original binary sequence
+    indices_of_array = []
+    for i in range(2**num_qubits):
+        indices_of_array.append(i)
+
+    # Check both indices could be indices for num_qubits array
+    try:
+        assert all(x in indices_of_array for x in [
+                index_of_state_1, index_of_state_2])
+    except AssertionError:
+        raise IndexError(
+            "At least one of the input indices is invalid.")
+
+    # get the basis vector strings
+    binary_sequence_list = _binary_sequence(num_qubits)
+    state_1_binary = binary_sequence_list[index_of_state_1]
+    state_2_binary = binary_sequence_list[index_of_state_2]
+
+    # compare binary sequence positions to gray code sequence
+    # finds number of steps neeeded to go from one state to another
+    gray_code_sequence = _gray_code_sequence(num_qubits)
+    state_1_gray_code_index = gray_code_sequence.index(state_1_binary)
+    state_2_gray_code_index = gray_code_sequence.index(state_2_binary)
+    num_steps_gray_code = state_2_gray_code_index - state_1_gray_code_index
+
+    # create a smaller gray code sequence between the two states of interest
+    gray_code = _gray_code_sequence(num_qubits)[
+        state_1_gray_code_index:state_2_gray_code_index+1]
+    return(gray_code, num_steps_gray_code)
+
+
+def gray_code_plot(index_of_state_1, index_of_state_2, num_qubits):
+    """ Plots the difference between each step of a gray code sequence.
+
+    Parameters
+    -----------
+    index_of_state_i: int
+        The non-trivial indices in a two-level unitary matrix gate.
+
+    num_qubits:
+        Number of qubits being acted upon by the quantum gate.
+    """
+    # This function is still incomplete.
+
+
+def gray_code_gate_info(index_of_state_1, index_of_state_2, num_qubits):
+    """ Finds information about control and targets for CNOT gate in a gray
+    code sequence.
+    """
+    gray_code_info = _gray_code_steps(
+        index_of_state_1, index_of_state_2, num_qubits)
+    gray_code = gray_code_info[0]
+    num_steps = gray_code_info[1]
+
+    # find number of ones in each gray code step, this will be used
+    # to keep track of control values.
+    one_bit_count = []
+    for i in range(num_steps+1):
+        check_bit = gray_code[i].count("1")
+        one_bit_count.append(check_bit)
+
+    gate_dictionary = {}
+    for i in range(num_steps):
+        a = gray_code[i]
+        b = gray_code[i+1]
+        for j in range(num_qubits):
+            gate_controls = []
+            gate_target = []
+            gate_control_value = []
+            while a[j] == b[j]:
+                partial_gate_control = j
+                gate_controls.append(partial_gate_control)
+                gate_control_value.append(a[j])
+            else:
+                partial_gate_target = j
+                gate_target.append(partial_gate_target)
+
+        gate_dictionary[i] = [gate_controls, gate_control_value, gate_target]
```

Comment:
@hodgestar @BoxiLi  This function fails to return an output because it gets stuck in an infinite loop.  Not sure what might be causing it.

Reply:
If it runs infinitely then there is likely a bug in the `while` loop.

Reply:
Could you provide an example input for this function?

Reply:
It might be wrong but in the `while a[j] == b[j]` loop you never change `a` and `b` so `a[j] == b[j]` is always true and the while loop never stops?

Reply:
> Could you provide an example input for this function?

Apologies for the delay. I have managed to correct my function definition. You should be able to see the code for below tomorrow morning (I haven't finished adding tests yet). 

Consider some example 3 qubit two-level unitary acting non-trivially on `010` and `101`.

The full binary sequence can be found by using  `_binary_sequence(3)` which will output `['000', '001', '010', '011', '100', '101', '110', '111']`

The full gray code sequence is `['000', '001', '011', '010', '110', '111', '101', '100']` by using ` _gray_code_sequence(3)` or in terms of index positions of binary sequence as `[0, 1, 3, 2, 6, 7, 5, 4]`. 

Thus, it can be seen that our examples indices `010` and `101` can be translated to binary indices as 2 and 5. The partial gray code sequence and the total number of steps can be found via `_gray_code_steps(2, 5, 3)` which outputs a tuple of the partial gray code sequence and number of steps. 

```
(['010', '110', '111', '101'], 3)
```

Now, our goal is to find the information about CNOT gates for mapping from  `010` to `101` via `gray_code_gate_info(2, 5, 3)`. 

The output of this function is in the form of a dictionary with key as iteration value. 
```
{0: {'controls =': [1, 2], 'control_value =': ['1', '0'], 'targets =': [0]}, 1: {'controls =': [0, 1], 'control_value =': ['1', '1'], 'targets =': [2]}, 2: {'controls =': [0, 2], 'control_value =': ['1', '1'], 'targets =': [1]}}
```